content/languages/ruby/authoring.md

```diff
@@ -0,0 +1,156 @@
+
+The reference solution or reference data ___must not___ be defined in the [Preloaded code](/authoring/guidelines/preloaded/).
+
+### Calling assertions
+
+The Ruby testing framework provides a set of useful [assertions](/languages/ruby/codewars-test/#assertions), but when used incorrectly, they can cause a series of problems:
+- Use of an assertion not suitable for the given case may lead to incorrect test results.
+- Incorrectly used assertions may produce confusing or unhelpful messages.
+
+To avoid the above problems, calls to assertion functions should respect the following rules:
+- The expected value should be calculated _before_ invoking an assertion. The `expected` parameter passed to the assertion should not be a lambda/`Proc`/block, but a value calculated directly beforehand.
+- Appropriate assertion functions should be used for each given test. `Test.assert_equals` is not suitable in all situations. Use `Test.expect` for tests on boolean values, and `Test.expect_error` to test error handling.
+<!-- Use `Test.assert_approx_equals` for floating-point comparisons  -->
+<!-- TODO: create snippet for `assert_approx_equals` equivalent in Ruby -->
+- Some additional attention should be paid to the order of parameters passed to assertion functions. It differs between various assertion libraries, and it happens to be quite often confused by authors, mixing up `actual` and `expected` in assertion messages. For the Ruby testing framework, the order is `(actual, expected)`.
+- One somewhat distinctive feature of Ruby assertions is that by default, a failed assertion does not cause a test case to fail early. It can lead to unexpected crashes when an actual value had already been asserted to be invalid, but the execution of the current test case was not stopped and following assertions continue to refer to it.
+- To avoid unexpected crashes in tests, it's recommended to perform some additional assertions before assuming that the answer returned by the user solution has some particular type, form, or value. For example, if the test suite sorts the returned array to verify its correctness, an explicit assertion should be added to check whether the returned object is really a list, and not, for example, `nil`.
+
+## Example test suite
+
+Below you can find an example test suite that covers most of the common scenarios mentioned in this article. Note that it does not present all possible techniques, so actual test suites can use a different structure, as long as they keep to established conventions and do not violate authoring guidelines.
+
+<!-- TODO -->
```

Comment:
about the example test suite, ruby is a total mess because files are concatenated (preloaded/solution/tests) so everything is in the same scope.
IIRC, it's even possible to access subfunctions, no? I don't actually remember (that mess is part of the reasons I totally dropped ruby), but I'm sure that making the ref solution inaccessible to the user is a hassle with ruby...

Reply:
I mentioned that on line 133. You just have to use something other than `def`.

Reply:
yeah, which leads to order definition troubles x) (assuming you're building lambdas)